// @tarojs/components-react/dist/original/components/button/index.js
import { __rest } from 'tslib';
import './style/index.scss';
import classNames from 'classnames';
import { omit, createForwardRefComponent } from '../../utils/index.js';
import { useRef, useState, useEffect, useCallback } from '../../utils/hooks.react.js';
import { jsxs, jsx } from 'react/jsx-runtime';

function Button(props) {
  const startTimer = useRef();
  const endTimer = useRef();
  const [state, setState] = useState({
    hover: false,
    touch: false
  });
  useEffect(() => {
    return () => {
      startTimer.current && clearTimeout(startTimer.current);
      endTimer.current && clearTimeout(endTimer.current);
    };
  }, []);
  const _onTouchStart = e => {
    setState(e => Object.assign(Object.assign({}, e), {
      touch: true
    }));
    if (props.hoverClass && props.hoverClass !== 'none' && !props.disabled) {
      startTimer.current = setTimeout(() => {
        if (state.touch) {
          setState(e => Object.assign(Object.assign({}, e), {
            hover: true
          }));
        }
      }, props.hoverStartTime || 20);
    }
    props.onTouchStart && props.onTouchStart(e);
  };
  const _onTouchEnd = e => {
    setState(e => Object.assign(Object.assign({}, e), {
      touch: false
    }));
    if (props.hoverClass && props.hoverClass !== 'none' && !props.disabled) {
      endTimer.current = setTimeout(() => {
        if (!state.touch) {
          setState(e => Object.assign(Object.assign({}, e), {
            hover: false
          }));
        }
      }, props.hoverStayTime || 70);
    }
    props.onTouchEnd && props.onTouchEnd(e);
  };
  const {
      forwardedRef,
      plain = false,
      children,
      disabled = false,
      className,
      style,
      onClick,
      hoverClass = 'button-hover',
      loading = false,
      type = 'default',
      size
    } = props,
    restProps = __rest(props, ["forwardedRef", "plain", "children", "disabled", "className", "style", "onClick", "hoverClass", "loading", "type", "size"]);
  const handleClick = useCallback(e => {
    if (disabled) return; // 如果按钮不可用，直接返回
    onClick === null || onClick === void 0 ? void 0 : onClick(e); // 否则执行点击回调
  }, [disabled, onClick]);
  const cls = classNames(className, 'taro-button-core', {
    [`${hoverClass}`]: state.hover && !disabled,
    'taro-btn-disabled': disabled,
    'taro-btn-loading': loading,
    'taro-btn-plain': plain,
    'taro-btn-mini': size === 'mini',
    'taro-btn-default': type === 'default',
    'taro-btn-primary': type === 'primary',
    'taro-btn-warn': type === 'warn'
  });
  return /*#__PURE__*/jsxs("button", {
    ...omit(restProps, ['hoverClass', 'onTouchStart', 'onTouchEnd', 'type', 'loading', 'forwardedRef', 'size', 'plain', 'disabled', 'onClick']),
    type: type,
    size: size,
    disabled: disabled,
    ref: forwardedRef,
    className: cls,
    style: style,
    onClick: handleClick,
    onTouchStart: _onTouchStart,
    onTouchEnd: _onTouchEnd,
    loading: loading.toString(),
    plain: plain.toString(),
    children: [!!loading && /*#__PURE__*/jsx("i", {
      className: "weui-loading"
    }), children]
  });
}
var index = createForwardRefComponent(Button);

export { index as default };
